### PR TITLE
cpu/msp430: improve periph_timer

### PR DESCRIPTION
**Pull request by maribu**
### Contribution description

- add support for multiple timers
- add support for selecting clock source in the board's `periph_conf.h`
- add support for the prescaler
- implement `periph_timer_query_freqs`
- add a second timer to all MSP430 boards
    - the first timer is fast ticking, high-power
    - the second is slow ticking, low-power

### Testing procedure

<details><summary><code>make BOARD=olimex-msp430-h1611 flash test -C tests/periph/timer</code></summary>

```
make: Entering directory '/home/maribu/Repos/software/RIOT/msp430/tests/periph/timer'
Building application "tests_timer" for "olimex-msp430-h1611" with MCU "msp430".

"make" -C /home/maribu/Repos/software/RIOT/msp430/boards/common/init
"make" -C /home/maribu/Repos/software/RIOT/msp430/boards/olimex-msp430-h1611
"make" -C /home/maribu/Repos/software/RIOT/msp430/core
"make" -C /home/maribu/Repos/software/RIOT/msp430/core/lib
"make" -C /home/maribu/Repos/software/RIOT/msp430/cpu/msp430
"make" -C /home/maribu/Repos/software/RIOT/msp430/cpu/msp430/periph
"make" -C /home/maribu/Repos/software/RIOT/msp430/drivers
"make" -C /home/maribu/Repos/software/RIOT/msp430/drivers/periph_common
"make" -C /home/maribu/Repos/software/RIOT/msp430/sys
"make" -C /home/maribu/Repos/software/RIOT/msp430/sys/auto_init
"make" -C /home/maribu/Repos/software/RIOT/msp430/sys/div
"make" -C /home/maribu/Repos/software/RIOT/msp430/sys/isrpipe
"make" -C /home/maribu/Repos/software/RIOT/msp430/sys/libc
"make" -C /home/maribu/Repos/software/RIOT/msp430/sys/malloc_thread_safe
"make" -C /home/maribu/Repos/software/RIOT/msp430/sys/newlib_syscalls_default
"make" -C /home/maribu/Repos/software/RIOT/msp430/sys/preprocessor
"make" -C /home/maribu/Repos/software/RIOT/msp430/sys/stdio_uart
"make" -C /home/maribu/Repos/software/RIOT/msp430/sys/test_utils/interactive_sync
"make" -C /home/maribu/Repos/software/RIOT/msp430/sys/test_utils/print_stack_usage
"make" -C /home/maribu/Repos/software/RIOT/msp430/sys/tsrb
/usr/lib/gcc/msp430-elf/13.1.0/../../../../msp430-elf/bin/ld: warning: /home/maribu/Repos/software/RIOT/msp430/tests/periph/timer/bin/olimex-msp430-h1611/tests_timer.elf has a LOAD segment with RWX permissions
   text	  data	   bss	   dec	   hex	filename
  11425	   180	  1186	 12791	  31f7	/home/maribu/Repos/software/RIOT/msp430/tests/periph/timer/bin/olimex-msp430-h1611/tests_timer.elf
mspdebug -j olimex "prog /home/maribu/Repos/software/RIOT/msp430/tests/periph/timer/bin/olimex-msp430-h1611/tests_timer.hex"
MSPDebug version 0.25 - debugging tool for MSP430 MCUs
Copyright (C) 2009-2017 Daniel Beer <dlbeer@gmail.com>
This is free software; see the source for copying conditions.  There is NO
warranty; not even for MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
Chip info database from MSP430.dll v3.3.1.4 Copyright (C) 2013 TI, Inc.

Resetting Olimex command processor...
Initializing FET...
FET protocol version is 20000007
Set Vcc: 3000 mV
Configured for JTAG (2)
Sending reset...
Using Olimex identification procedure
Device ID: 0xf16c
  Code start address: 0x4000
  Code size         : 49152 byte = 48 kb
  RAM  start address: 0x200
  RAM  end   address: 0x9ff
  RAM  size         : 2048 byte = 2 kb
Device: MSP430F1611
Number of breakpoints: 8
fet: FET returned error code 34 (Not supported by selected interface or interface is not initialized)
fet: warning: message C_IDENT3 failed
fet: FET returned NAK
warning: device does not support power profiling
Chip ID data:
  ver_id:         6cf1
  ver_sub_id:     0000
  revision:       20
  fab:            40
  self:           0000
  config:         00
  fuses:          00
Device: MSP430F1611
Erasing...
Programming...
Writing 1345 bytes at 4000...
Writing 4096 bytes at 4542...
Writing 4096 bytes at 5542...
Writing 2056 bytes at 6542...
Writing    2 bytes at ffe6...
Writing    4 bytes at ffea...
Writing    4 bytes at fff8...
Writing    2 bytes at fffe...
Done, 11605 bytes total
r
/home/maribu/Repos/software/RIOT/msp430/dist/tools/pyterm/pyterm -p "/dev/ttyUSB0" -b "115200" --no-reconnect --noprefix --no-repeat-command-on-empty-line 
Connect to serial port /dev/ttyUSB0
Welcome to pyterm!
Type '/exit' to exit.
READY
s
START
main(): This is RIOT! (Version: 2023.10-devel-683-g151c9-cpu/msp430/timer)

Test for peripheral TIMERs

Available timers: 2

TIMER 0
=======

  - supported frequencies:
    0: 5025792
    1: 2512896
    2: 1256448
    ....
    3: 628224
  - Calling timer_init(0, 5025792)
    initialization successful
  - timer_stop(0): stopped
  - timer_set(0, 0, 5000)
    Successfully set timeout 5000 for channel 0
  - timer_set(0, 1, 10000)
    Successfully set timeout 10000 for channel 1
  - timer_set(0, 2, 15000)
    Successfully set timeout 15000 for channel 2
  - timer_start(0)
  - Results:
    - channel 0 fired at SW count      114      - init:      114
    - channel 1 fired at SW count      224      - diff:      110
    - channel 2 fired at SW count      332      - diff:      108
  - Validating no spurious IRQs are triggered:
    OK (no spurious IRQs)
  - Calling timer_init(0, 2512896)
    initialization successful
  - timer_stop(0): stopped
  - timer_set(0, 0, 5000)
    Successfully set timeout 5000 for channel 0
  - timer_set(0, 1, 10000)
    Successfully set timeout 10000 for channel 1
  - timer_set(0, 2, 15000)
    Successfully set timeout 15000 for channel 2
  - timer_start(0)
  - Results:
    - channel 0 fired at SW count      228      - init:      228
    - channel 1 fired at SW count      451      - diff:      223
    - channel 2 fired at SW count      673      - diff:      222
  - Validating no spurious IRQs are triggered:
    OK (no spurious IRQs)
  - Calling timer_init(0, 1256448)
    initialization successful
  - timer_stop(0): stopped
  - timer_set(0, 0, 5000)
    Successfully set timeout 5000 for channel 0
  - timer_set(0, 1, 10000)
    Successfully set timeout 10000 for channel 1
  - timer_set(0, 2, 15000)
    Successfully set timeout 15000 for channel 2
  - timer_start(0)
  - Results:
    - channel 0 fired at SW count      455      - init:      455
    - channel 1 fired at SW count      906      - diff:      451
    - channel 2 fired at SW count     1355      - diff:      449
  - Validating no spurious IRQs are triggered:
    OK (no spurious IRQs)

TIMER 1
=======

  - supported frequencies:
    0: 32768
    1: 16384
    2: 8192
    ....
    3: 4096
  - Calling timer_init(1, 32768)
    initialization successful
  - timer_stop(1): stopped
  - timer_set(1, 0, 5000)
    Successfully set timeout 5000 for channel 0
  - timer_set(1, 1, 10000)
    Successfully set timeout 10000 for channel 1
  - timer_set(1, 2, 15000)
    Successfully set timeout 15000 for channel 2
  - timer_set(1, 3, 20000)
    Successfully set timeout 20000 for channel 3
  - timer_set(1, 4, 25000)
    Successfully set timeout 25000 for channel 4
  - timer_set(1, 5, 30000)
    Successfully set timeout 30000 for channel 5
  - timer_set(1, 6, 35000)
    Successfully set timeout 35000 for channel 6
  - timer_start(1)
  - Results:
    - channel 0 fired at SW count    17428      - init:    17428
    - channel 1 fired at SW count    34852      - diff:    17424
    - channel 2 fired at SW count    52275      - diff:    17423
    - channel 3 fired at SW count     4161      - diff:    17422
    - channel 4 fired at SW count    21583      - diff:    17422
    - channel 5 fired at SW count    39005      - diff:    17422
    - channel 6 fired at SW count    56427      - diff:    17422
  - Validating no spurious IRQs are triggered:
    OK (no spurious IRQs)
  - Calling timer_init(1, 16384)
    initialization successful
  - timer_stop(1): stopped
  - timer_set(1, 0, 5000)
    Successfully set timeout 5000 for channel 0
  - timer_set(1, 1, 10000)
    Successfully set timeout 10000 for channel 1
  - timer_set(1, 2, 15000)
    Successfully set timeout 15000 for channel 2
  - timer_set(1, 3, 20000)
    Successfully set timeout 20000 for channel 3
  - timer_set(1, 4, 25000)
    Successfully set timeout 25000 for channel 4
  - timer_set(1, 5, 30000)
    Successfully set timeout 30000 for channel 5
  - timer_set(1, 6, 35000)
    Successfully set timeout 35000 for channel 6
  - timer_start(1)
  - Results:
    - channel 0 fired at SW count    34849      - init:    34849
    - channel 1 fired at SW count     4164      - diff:    34851
    - channel 2 fired at SW count    39014      - diff:    34850
    - channel 3 fired at SW count     8327      - diff:    34849
    - channel 4 fired at SW count    43176      - diff:    34849
    - channel 5 fired at SW count    12491      - diff:    34851
    - channel 6 fired at SW count    47340      - diff:    34849
  - Validating no spurious IRQs are triggered:
    OK (no spurious IRQs)
  - Calling timer_init(1, 8192)
    initialization successful
  - timer_stop(1): stopped
  - timer_set(1, 0, 5000)
    Successfully set timeout 5000 for channel 0
  - timer_set(1, 1, 10000)
    Successfully set timeout 10000 for channel 1
  - timer_set(1, 2, 15000)
    Successfully set timeout 15000 for channel 2
  - timer_set(1, 3, 20000)
    Successfully set timeout 20000 for channel 3
  - timer_set(1, 4, 25000)
    Successfully set timeout 25000 for channel 4
  - timer_set(1, 5, 30000)
    Successfully set timeout 30000 for channel 5
  - timer_set(1, 6, 35000)
    Successfully set timeout 35000 for channel 6
  - timer_start(1)
  - Results:
    - channel 0 fired at SW count     4170      - init:    69706
    - channel 1 fired at SW count     8347      - diff:    69713
    - channel 2 fired at SW count    12522      - diff:    69711
    - channel 3 fired at SW count    16697      - diff:    69711
    - channel 4 fired at SW count    20871      - diff:    69710
    - channel 5 fired at SW count    25044      - diff:    69709
    - channel 6 fired at SW count    29218      - diff:    69710
  - Validating no spurious IRQs are triggered:
    OK (no spurious IRQs)

TEST SUCCEEDED

make: Leaving directory '/home/maribu/Repos/software/RIOT/msp430/tests/periph/timer'
```

</details>

### Issues/PRs references

Implements: https://github.com/RIOT-OS/RIOT/pull/16349
Depends on and includes:
- [ ] https://github.com/RIOT-OS/RIOT/pull/20161

### CODE REVIEW

**Comment by maribu** [comment]
> Could this be tested on a different MSP430 board as well?

Yes, any MSP430 board supported by RIOT has the same two timer peripherals. Only the frequencies the first timer can generate will differ when the CPU clock is different.

Note that the tests for the 32.678 kHz timer is taking a couple of seconds due to the slow speed of the timer and the test iterating over three different frequencies the timer claims to support.

**Comment by maribu** [comment]
Looks like there is still an issue with spurious IRQs on MSP430 F2xx/G2xx:

<details><summary><code>make BOARD=olimex-msp430-h2618 flash test -j -C tests/periph/timer</code></summary>

```
make: Entering directory '/home/maribu/Repos/software/RIOT/master/tests/periph/timer'
Building application "tests_timer" for "olimex-msp430-h2618" with MCU "msp430".

"make" -C /home/maribu/Repos/software/RIOT/master/boards/common/init
"make" -C /home/maribu/Repos/software/RIOT/master/boards/olimex-msp430-h2618
"make" -C /home/maribu/Repos/software/RIOT/master/core
"make" -C /home/maribu/Repos/software/RIOT/master/core/lib
"make" -C /home/maribu/Repos/software/RIOT/master/cpu/msp430
"make" -C /home/maribu/Repos/software/RIOT/master/drivers
"make" -C /home/maribu/Repos/software/RIOT/master/sys
"make" -C /home/maribu/Repos/software/RIOT/master/drivers/periph_common
"make" -C /home/maribu/Repos/software/RIOT/master/sys/auto_init
"make" -C /home/maribu/Repos/software/RIOT/master/sys/div
"make" -C /home/maribu/Repos/software/RIOT/master/cpu/msp430/periph
"make" -C /home/maribu/Repos/software/RIOT/master/sys/isrpipe
"make" -C /home/maribu/Repos/software/RIOT/master/sys/libc
"make" -C /home/maribu/Repos/software/RIOT/master/sys/malloc_thread_safe
"make" -C /home/maribu/Repos/software/RIOT/master/sys/newlib_syscalls_default
"make" -C /home/maribu/Repos/software/RIOT/master/sys/preprocessor
"make" -C /home/maribu/Repos/software/RIOT/master/sys/stdio_uart
"make" -C /home/maribu/Repos/software/RIOT/master/sys/test_utils/interactive_sync
"make" -C /home/maribu/Repos/software/RIOT/master/sys/test_utils/print_stack_usage
"make" -C /home/maribu/Repos/software/RIOT/master/sys/tsrb
/usr/lib/gcc/msp430-elf/13.1.0/../../../../msp430-elf/bin/ld: warning: /home/maribu/Repos/software/RIOT/master/tests/periph/timer/bin/olimex-msp430-h2618/tests_timer.elf has a LOAD segment with RWX permissions
   text	  data	   bss	   dec	   hex	filename
  11115	   168	  1186	 12469	  30b5	/home/maribu/Repos/software/RIOT/master/tests/periph/timer/bin/olimex-msp430-h2618/tests_timer.elf
mspdebug -j olimex "prog /home/maribu/Repos/software/RIOT/master/tests/periph/timer/bin/olimex-msp430-h2618/tests_timer.hex"
MSPDebug version 0.25 - debugging tool for MSP430 MCUs
Copyright (C) 2009-2017 Daniel Beer <dlbeer@gmail.com>
This is free software; see the source for copying conditions.  There is NO
warranty; not even for MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
Chip info database from MSP430.dll v3.3.1.4 Copyright (C) 2013 TI, Inc.

Resetting Olimex command processor...
Initializing FET...
FET protocol version is 20000007
Set Vcc: 3000 mV
Configured for JTAG (2)
Sending reset...
Using Olimex identification procedure
Device ID: 0xf26f
  Code start address: 0x3100
  Code size         : 118528 byte = 115 kb
  RAM  start address: 0x200
  RAM  end   address: 0x9ff
  RAM  size         : 2048 byte = 2 kb
Device: MSP430F2618
Number of breakpoints: 8
fet: FET returned error code 34 (Not supported by selected interface or interface is not initialized)
fet: warning: message C_IDENT3 failed
fet: FET returned NAK
warning: device does not support power profiling
Chip ID data:
  ver_id:         6ff2
  ver_sub_id:     0000
  revision:       07
  fab:            60
  self:           0000
  config:         00
  fuses:          00
Device: MSP430F2619
Erasing...
Programming...
Writing 1345 bytes at 3100...
Writing 4096 bytes at 3642...
Writing 4096 bytes at 4642...
Writing 1734 bytes at 5642...
Writing    6 bytes at ffee...
Writing    4 bytes at fff8...
Writing    2 bytes at fffe...
Done, 11283 bytes total
Connect to serial port /dev/ttyUSB0
Welcome to pyterm!
Type '/exit' to exit.
Help: Press s to start test, r to print it is ready
READY
s
START
main(): This is RIOT! (Version: 2023.10-devel-685-g81acb-cpu/msp430/timer)

Test for peripheral TIMERs

Available timers: 2

TIMER 0
=======

  - supported frequencies:
    0: 16027648
    1: 8013824
    2: 4006912
    ....
    3: 2003456
  - Calling timer_init(0, 16027648)
    initialization successful
  - timer_stop(0): stopped
  - timer_set(0, 0, 5000)
    Successfully set timeout 5000 for channel 0
  - timer_set(0, 1, 10000)
    Successfully set timeout 10000 for channel 1
  - timer_set(0, 2, 15000)
    Successfully set timeout 15000 for channel 2
  - timer_start(0)
  - Results:
    - channel 0 fired at SW count      125      - init:      125
    - channel 1 fired at SW count      246      - diff:      121
    - channel 2 fired at SW count      366      - diff:      120
  - Validating no spurious IRQs are triggered:
    ERROR: Spurious timer fired (2/2)
  - Calling timer_init(0, 8013824)
    initialization successful
  - timer_stop(0): stopped
  - timer_set(0, 0, 5000)
    Successfully set timeout 5000 for channel 0
  - timer_set(0, 1, 10000)
    Successfully set timeout 10000 for channel 1
  - timer_set(0, 2, 15000)
    Successfully set timeout 15000 for channel 2
  - timer_start(0)
  - Results:
    - channel 0 fired at SW count      250      - init:      250
    - channel 1 fired at SW count      496      - diff:      246
    - channel 2 fired at SW count      741      - diff:      245
  - Validating no spurious IRQs are triggered:
    OK (no spurious IRQs)
  - Calling timer_init(0, 4006912)
    initialization successful
  - timer_stop(0): stopped
  - timer_set(0, 0, 5000)
    Successfully set timeout 5000 for channel 0
  - timer_set(0, 1, 10000)
    Successfully set timeout 10000 for channel 1
  - timer_set(0, 2, 15000)
    Successfully set timeout 15000 for channel 2
  - timer_start(0)
  - Results:
    - channel 0 fired at SW count      500      - init:      500
    - channel 1 fired at SW count      996      - diff:      496
    - channel 2 fired at SW count     1491      - diff:      495
  - Validating no spurious IRQs are triggered:
    OK (no spurious IRQs)

TIMER 1
=======

  - supported frequencies:
    0: 32768
    1: 16384
    2: 8192
    ....
    3: 4096
  - Calling timer_init(1, 32768)
    initialization successful
  - timer_stop(1): stopped
  - timer_set(1, 0, 5000)
    Successfully set timeout 5000 for channel 0
  - timer_set(1, 1, 10000)
    Successfully set timeout 10000 for channel 1
  - timer_set(1, 2, 15000)
    Successfully set timeout 15000 for channel 2
  - timer_set(1, 3, 20000)
    Successfully set timeout 20000 for channel 3
  - timer_set(1, 4, 25000)
    Successfully set timeout 25000 for channel 4
  - timer_set(1, 5, 30000)
    Successfully set timeout 30000 for channel 5
  - timer_set(1, 6, 35000)
    Successfully set timeout 35000 for channel 6
  - timer_start(1)
  - Results:
    - channel 0 fired at SW count    61112      - init:    61112
    - channel 1 fired at SW count    56691      - diff:    61115
    - channel 2 fired at SW count    52268      - diff:    61113
    - channel 3 fired at SW count    47849      - diff:    61117
    - channel 4 fired at SW count    43420      - diff:    61107
    - channel 5 fired at SW count    38987      - diff:    61103
    - channel 6 fired at SW count    34561      - diff:    61110
  - Validating no spurious IRQs are triggered:
    OK (no spurious IRQs)
  - Calling timer_init(1, 16384)
    initialization successful
  - timer_stop(1): stopped
  - timer_set(1, 0, 5000)
    Successfully set timeout 5000 for channel 0
  - timer_set(1, 1, 10000)
    Successfully set timeout 10000 for channel 1
  - timer_set(1, 2, 15000)
    Successfully set timeout 15000 for channel 2
  - timer_set(1, 3, 20000)
    Successfully set timeout 20000 for channel 3
  - timer_set(1, 4, 25000)
    Successfully set timeout 25000 for channel 4
  - timer_set(1, 5, 30000)
    Successfully set timeout 30000 for channel 5
  - timer_set(1, 6, 35000)
    Successfully set timeout 35000 for channel 6
  - timer_start(1)
  - Results:
    - channel 0 fired at SW count    56665      - init:   122201
    - channel 1 fired at SW count    47807      - diff:   122214
    - channel 2 fired at SW count    38944      - diff:   122209
    - channel 3 fired at SW count    30083      - diff:   122211
    - channel 4 fired at SW count    21228      - diff:   122217
    - channel 5 fired at SW count    12364      - diff:   122208
    - channel 6 fired at SW count     3529      - diff:   122237
  - Validating no spurious IRQs are triggered:
    OK (no spurious IRQs)
  - Calling timer_init(1, 8192)
    initialization successful
  - timer_stop(1): stopped
  - timer_set(1, 0, 5000)
    Successfully set timeout 5000 for channel 0
  - timer_set(1, 1, 10000)
    Successfully set timeout 10000 for channel 1
  - timer_set(1, 2, 15000)
    Successfully set timeout 15000 for channel 2
  - timer_set(1, 3, 20000)
    Successfully set timeout 20000 for channel 3
  - timer_set(1, 4, 25000)
    Successfully set timeout 25000 for channel 4
  - timer_set(1, 5, 30000)
    Successfully set timeout 30000 for channel 5
  - timer_set(1, 6, 35000)
    Successfully set timeout 35000 for channel 6
  - timer_start(1)
  - Results:
    - channel 0 fired at SW count    47818      - init:   244426
    - channel 1 fired at SW count    30107      - diff:   244433
    - channel 2 fired at SW count    12376      - diff:   244413
    - channel 3 fired at SW count    60213      - diff:   244445
    - channel 4 fired at SW count    42503      - diff:   244434
    - channel 5 fired at SW count    24805      - diff:   244446
    - channel 6 fired at SW count     7100      - diff:   244439
  - Validating no spurious IRQs are triggered:
    OK (no spurious IRQs)

TEST FAILED
```

</details>

**Comment by maribu** [comment]
> Looks like there is still an issue with spurious IRQs on MSP430 F2xx/G2xx:

No, the test was just buggy :seat_smile:

https://github.com/RIOT-OS/RIOT/pull/20161 fixes the issue.

File: boards/common/msp430/include/cfg_timer_a_smclk_b_aclk.h [hunk 1]
```diff
@@ -0,0 +1,61 @@
+/*
+ * Copyright (C) 2019Otto-von-Guericke-Universität Magdeburg
```
Comment (benpicco) [inline boards/common/msp430/include/cfg_timer_a_smclk_b_aclk.h:2]:
```suggestion
 * Copyright (C) 2023 Otto-von-Guericke-Universität Magdeburg
```

**Review by benpicco — APPROVED** [review]
I don't have the hardware, but I trust your testing.
Code wise this looks good and is certainly an improvement over the status quo.

**Comment by maribu** [comment]
`examples/gnrc_networking_minimal` still builds with my local toolchain and is a whopping 4 B larger in terms of RAM. (`.text` grew a few 100 B, but that is expected with an actual timer implementation now present.) Anyway, with the CI toolchain I need to touch the `Makefile.ci`s.

**Comment by maribu** [comment]
:tada: Thanks a lot! :-)

**Pull request merged**